app/multitenant/memcache_client.go

```diff
@@ -0,0 +1,131 @@
+	}, []string{"method", "status_code"})
+)
+
+func init() {
+	prometheus.MustRegister(memcacheCounter)
+	prometheus.MustRegister(memcacheRequestDuration)
+}
+
+type memcacheClient struct {
+	client     *memcache.Client
+	expiration int32
+}
+
+// Create a new MemcacheClient that gets its server list from SRV and updates
+// the server list on a regular basis.
+func NewMemcacheClient(host string, timeout time.Duration, service string, updateInterval time.Duration, expiration int32) (*memcacheClient, error) {
+	var servers memcache.ServerList
+	client := memcache.NewFromSelector(&servers)
+	client.Timeout = timeout
+	err := updateMemcacheServers(host, service, &servers)
+	if err != nil {
+		return nil, err
+	}
+
+	ticker := time.NewTicker(updateInterval)
+	go func() {
+		for range ticker.C {
+			updateMemcacheServers(host, service, &servers)
+		}
+	}()
+	return &memcacheClient{client, expiration}, nil
+}
+
+// updateMemcacheServers sets a memcache server list from SRV records. SRV
+// priority & weight are ignored.
+func updateMemcacheServers(hostname, service string, serverList *memcache.ServerList) error {
+	_, addrs, err := net.LookupSRV(service, "tcp", hostname)
+	if err != nil {
+		return err
+	}
+	var servers []string
+	for _, srv := range addrs {
+		servers = append(servers, fmt.Sprintf("%s:%d", srv.Target, srv.Port))
+	}
+	// ServerList deterministically maps keys to _index_ of the server list.
+	// Since DNS returns records in different order each time, we sort to
+	// guarantee best possible match between nodes.
+	sort.Strings(servers)
+	return serverList.SetServers(servers...)
+}
+
+func (c memcacheClient) FetchReports(keys []string) ([]report.Report, []string, error) {
+	var found map[string]*memcache.Item
+	err := timeRequestStatus("Get", memcacheRequestDuration, memcacheStatusCode, func() error {
+		var err error
+		found, err = c.client.GetMulti(keys)
+		return err
+	})
+	if err != nil {
+		return nil, keys, err
+	}
+
+	// Decode all the reports in parallel.
+	type result struct {
+		key    string
+		report *report.Report
+	}
+	ch := make(chan result, len(keys))
+	var missing []string
+	for _, key := range keys {
+		item, ok := found[key]
+		if !ok {
+			missing = append(missing, key)
+			continue
+		}
+		go func(key string) {
+			rep, err := report.MakeFromBinary(bytes.NewReader(item.Value))
+			if err != nil {
+				log.Warningf("Corrupt report in memcache %v: %v", key, err)
+				ch <- result{key: key}
+				return
+			}
+			ch <- result{report: rep}
+		}(key)
+	}
+
+	var reports []report.Report
+	for i := 0; i < len(keys)-len(missing); i++ {
+		r := <-ch
+		if r.report == nil {
+			missing = append(missing, r.key)
+		} else {
+			reports = append(reports, *r.report)
+		}
+	}
+
+	memcacheCounter.WithLabelValues(hitLabel).Add(float64(len(reports)))
+	memcacheCounter.WithLabelValues(missLabel).Add(float64(len(missing)))
+	return reports, missing, nil
+}
+
+func (c memcacheClient) StoreBytes(key string, content []byte) error {
```

Comment:
I _think_ it's OK as is (the pointer to `client` gets copied and then `Set` is called on the original), but I've changed it because thinking about things sucks.